internal/controller/linodefirewall_controller_helpers.go

```diff
@@ -311,13 +314,53 @@ func processAddresses(addresses *infrav1alpha2.NetworkAddresses) (ipv4s, ipv6s [
 	return slices.Collect(maps.Keys(ipv4Set)), slices.Collect(maps.Keys(ipv6Set))
 }
 
-// get finalizer for Firewall that references the AddressSet
+// get finalizer for Firewall that references the AddressSet or FirewallRule
 func getFinalizer(lfw *infrav1alpha2.LinodeFirewall) string {
 	return fmt.Sprintf("lfw.%s.%s", lfw.Namespace, lfw.Name)
 }
 
+func addAddrSetFinalizer(ctx context.Context, k8sClient clients.K8sClient, firewall *infrav1alpha2.LinodeFirewall, addrSetRef *corev1.ObjectReference) (*infrav1alpha2.AddressSet, error) {
+	addrSet := &infrav1alpha2.AddressSet{}
+	if addrSetRef.Namespace == "" {
+		addrSetRef.Namespace = firewall.Namespace
+	}
+	finalizer := getFinalizer(firewall)
```

Comment:
If we delete a `FirewallRule` reference in the `LinodeFirewall`'s spec, does the `LinodeFirewall`'s finalizer get cleaned up?

Reply:
Short answer is no. Longer answer is there’s no way AFAICT to act on an outdated spec to fetch the resource and remove its finalizer on an update. There might be a way to check with an updateFunc predicate if refs were removed and remove the finalizer(s) but that’s going to have to ignore any errors it runs into because the UpdateFunc is just supposed to return a boolean if a reconcile needs to happen, it really shouldn’t be doing resource updating.
It might be preferable to remove the finalizers altogether on the dependent resources if there's no reasonable way to get the older resource version in the actual reconcile loop to remove the finalizer once it's de-referenced from the LinodeFirewall